File: frontend/src/api/workspace.ts
import { get, post, put, del } from './client';
import type { Workspace } from '../types';

export const createWorkspace = (data: { name: string; description?: string }) =>
  post<Workspace>('/workspaces', data);

export const listWorkspaces = () => get<Workspace[]>('/workspaces');

export const getWorkspace = (id: string) => get<Workspace>(`/workspace/${id}`);

export const updateWorkspace = (id: string, data: { name?: string; description?: string }) =>
  put<Workspace>(`/workspace/${id}`, data);

export const deleteWorkspace = (id: string) => del<{ deleted: boolean }>(`/workspace/${id}`);

File: backend/src/services/workspace.rs
use crate::errors::AppError;
use crate::models::workspace::Workspace;
use crate::repositories::neo4j_repo::Neo4jRepo;

pub struct WorkspaceService;

impl WorkspaceService {
    pub async fn create(repo: &Neo4jRepo, name: String, description: Option<String>) -> Result<Workspace, AppError> {
        let id = uuid::Uuid::new_v4().to_string();
        let created_at = chrono::Utc::now().to_rfc3339();
        let desc = description.unwrap_or_default();
        repo.create_workspace(&id, &name, &desc, &created_at).await
    }

    pub async fn list(repo: &Neo4jRepo) -> Result<Vec<Workspace>, AppError> {
        repo.list_workspaces().await
    }

    pub async fn get(repo: &Neo4jRepo, id: &str) -> Result<Workspace, AppError> {
        repo.get_workspace(id)
            .await?
            .ok_or_else(|| AppError::WorkspaceNotFound(id.to_string()))
    }

    pub async fn update(repo: &Neo4jRepo, id: &str, name: Option<String>, description: Option<String>) -> Result<Workspace, AppError> {
        repo.update_workspace(id, name.as_deref(), description.as_deref())
            .await?
            .ok_or_else(|| AppError::WorkspaceNotFound(id.to_string()))
    }

    pub async fn delete(repo: &Neo4jRepo, id: &str) -> Result<(), AppError> {
        let deleted = repo.delete_workspace(id).await?;
        if !deleted {
            return Err(AppError::WorkspaceNotFound(id.to_string()));
        }
        Ok(())
    }
}
